lib/token-store.js

```diff
@@ -8,196 +8,1110 @@
+
+        return tokens;
+    }
 
     /**
-     * Gets tokens in a given interval.
-     * @param {int} start Inclusive index of the first token. 0 if negative.
-     * @param {int} end Exclusive index of the last token.
-     * @returns {Token[]} Tokens in the interval.
+     * Gets the last tokens.
+     * This consumes this cursor.
+     * @type {Token|Comment}
      */
-    function get(start, end) {
-        const result = [];
+    get lastTokens() {
+        const tokens = [];
 
-        for (let i = Math.max(0, start); i < end && i < length; i++) {
-            result.push(tokens[i]);
+        while (this.movePrev()) {
+            tokens.push(this.current);
         }
+        tokens.reverse();
 
-        return result;
+        return tokens;
     }
 
     /**
-     * Gets the index in the tokens array of the last token belonging to a node.
-     * Usually a node ends exactly at a token, but due to ASI, sometimes a
-     * node's range extends beyond its last token.
-     * @param {ASTNode} node The node for which to find the last token's index.
-     * @returns {int} Index in the tokens array of the node's last token.
+     * Moves this cursor to the next token.
+     * @returns {boolean} `true` if the next token exists.
+     * @abstract
      */
-    function lastTokenIndex(node) {
-        const end = node.range[1];
-        let cursor = ends[end];
+    /* istanbul ignore next */
+    moveNext() { // eslint-disable-line class-methods-use-this
+        throw new Error("Not implemented.");
+    }
+
+    /**
+     * Moves this cursor to the previous token.
+     * @returns {boolean} `true` if the previous token exists.
+     * @abstract
+     */
+    /* istanbul ignore next */
+    movePrev() { // eslint-disable-line class-methods-use-this
+        throw new Error("Not implemented.");
+    }
+}
+
+/**
+ * The abstract class about cursors which iterate tokens.
+ * This is the specific class to directly iterate tokens from tokens' array.
+ *
+ * @private
+ */
+class BaseCursor extends Cursor {
+
+    /**
+     * Initializes this cursor.
+     * @param {Token[]} tokens - The array of tokens.
+     * @param {Comment[]} comments - The array of comments.
+     * @param {Object} imap - The map from locations to indices in `this.tokens`.
+     * @param {number} start - The start location of the iteration range.
+     * @param {number} end - The end location of the iteration range.
+     */
+    constructor(tokens, comments, imap, start, end) {
+        super();
+        this.tokens = tokens;
+        this.comments = comments;
+        this.imap = imap;
+        this.start = start | 0;
+        this.end = end | 0;
+    }
+
+    /**
+     * Gets the index of the location `this.start` in `this.tokens`.
+     * `this.start` can be the value of `node.range[1]`, so this checks about `this.start - 1` as well.
+     * @type {number}
+     * @protected
+     */
+    get firstIndex() {
+        if (typeof this.imap[this.start] !== "undefined") {
```

Comment:
I will fix it.
I thought `in` operator is slow since I had seen nodejs/node#10443. But this seems a different situation.
https://jsperf.com/undefined-check-777